### DISCUSSION TITLE
Directive constructor() parameters to include arguments also?

### DISCUSSION THREAD

**Discussion by miksula** [lit/lit#2212] [Ideas]
Is there specific reason why Directive constructor is only passed the part, but no props?

I have a scenario where I would like to access directive arguments during one time setup
process with the constructor(). 

Example: 
```
class MyDirective extends Directive {
  constructor(partInfo, [helloMessage]) {
    super(partInfo);
    console.log('MyDirective message', helloMessage);
  }
}

const myDirective = directive(MyDirective);

const template = html`<div>${myDirective("Hello!")}</div>`;
```

**Comment by justinfagnani** [comment]
Since the constructor is only called once, the constructor is only provided the things that _can't_ change per-render about a single directive instance, which is the PartInfo. The directive parameters can change every render, so they're only provided in `update()` and `render()`.

What are you trying to do that you want constructor parameters?

**Reply by miksula** [reply]
Thank you for the clarification, makes sense. I just understood that it is pretty straightforward to implement this "only once updated" logic with a custom directive (taking inspiration from the `ref` directive) that takes function as an argument:

```
class MountDirective extends Directive {
  _updated;
  update(part, [fn = (el, data) => {}, data = {}]) {
    if (this._updated) {
      return nothing;
    }
    fn(part.element, data)
    this._updated = true;
  }
  render() {}
}

const useEffectOnce = directive(MountDirective);
````

My use case is to use only lit-html for functional rendering. I needed React style useEffect() hook, that runs only once after the DOM element is available, and add event listers and use other DOM APIs on it.
